diff --git a/swift/test/Misc/verify-swift-feature-testing.test-sh b/swift/test/Misc/verify-swift-feature-testing.test-sh
index 3f36fc979cb..8713ae8fea8 100755
--- a/swift/test/Misc/verify-swift-feature-testing.test-sh
+++ b/swift/test/Misc/verify-swift-feature-testing.test-sh
@@ -98,7 +98,6 @@ def find_matches(swift_src_root):
     # `-enable-upcoming-feature` in the test directories.
     output = subprocess.check_output(
         [
-            "git",
             "grep",
             "--extended-regexp",
             "--recursive",
diff --git a/swift/utils/swift_build_support/swift_build_support/cmake.py b/swift/utils/swift_build_support/swift_build_support/cmake.py
index 9a0f5b43e2d..5c7d1320a2d 100644
--- a/swift/utils/swift_build_support/swift_build_support/cmake.py
+++ b/swift/utils/swift_build_support/swift_build_support/cmake.py
@@ -150,11 +150,8 @@ class CMake(object):
             define("CMAKE_CXX_COMPILER_LAUNCHER:PATH", args.cmake_cxx_launcher)
 
         if self.prefer_native_toolchain and product:
-            clang_tools_path = product.native_clang_tools_path(args.host_target)
-            define("CMAKE_C_COMPILER:PATH", os.path.join(clang_tools_path,
-                                                         'bin', 'clang'))
-            define("CMAKE_CXX_COMPILER:PATH", os.path.join(clang_tools_path,
-                                                           'bin', 'clang++'))
+            define("CMAKE_C_COMPILER:PATH", toolchain.cc)
+            define("CMAKE_CXX_COMPILER:PATH", toolchain.cxx)
 
             toolchain_path = product.native_toolchain_path(args.host_target)
             cmake_swiftc_path = os.getenv('CMAKE_Swift_COMPILER',
diff --git a/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake b/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake
index 3da519e..e2dd8ed 100644
--- a/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake
+++ b/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake
@@ -4,6 +4,10 @@ if(NOT dispatch_MODULE_TRIPLE)
   set(module_triple_command "${CMAKE_Swift_COMPILER}" -print-target-info)
   if(CMAKE_Swift_COMPILER_TARGET)
     list(APPEND module_triple_command -target ${CMAKE_Swift_COMPILER_TARGET})
+  else()
+    set(arg_list ${CMAKE_Swift_FLAGS})
+    separate_arguments(arg_list)
+    list(APPEND module_triple_command ${arg_list})
   endif()
   execute_process(COMMAND ${module_triple_command} OUTPUT_VARIABLE target_info_json)
 
diff --git a/swift-corelibs-libdispatch/src/swift/CMakeLists.txt b/swift-corelibs-libdispatch/src/swift/CMakeLists.txt
index 38bef37..d0ddf98 100644
--- a/swift-corelibs-libdispatch/src/swift/CMakeLists.txt
+++ b/swift-corelibs-libdispatch/src/swift/CMakeLists.txt
@@ -15,6 +15,7 @@ add_library(swiftDispatch
   Source.swift
   Time.swift
   Wrapper.swift)
+set(CMAKE_SHARED_LINKER_FLAGS "")
 target_compile_options(swiftDispatch PRIVATE
   "SHELL:-Xcc -fblocks"
   "SHELL:-Xcc -fmodule-map-file=${PROJECT_SOURCE_DIR}/dispatch/module.modulemap"
