diff --git a/swift/test/SILOptimizer/concat_string_literals.32.swift b/swift/test/SILOptimizer/concat_string_literals.32.swift
index 01390761c8c..d84c142397c 100644
--- a/swift/test/SILOptimizer/concat_string_literals.32.swift
+++ b/swift/test/SILOptimizer/concat_string_literals.32.swift
@@ -4,6 +4,7 @@
 
 // We have a separate test for 64-bit architectures.
 // REQUIRES: PTRSIZE=32
+// XFAIL: OS=linux-androideabi
 
 // NOTE: 25185.byteSwapped = 0x62 'a', 0x61 'b'
 // CHECK-LABEL: test_ascii_scalar_scalar2
diff --git a/swift/test/SILOptimizer/semantic-arc-opts-redundantcopyopts.sil b/swift/test/SILOptimizer/semantic-arc-opts-redundantcopyopts.sil
index 98765e0e4a3..4dc68bf522b 100644
--- a/swift/test/SILOptimizer/semantic-arc-opts-redundantcopyopts.sil
+++ b/swift/test/SILOptimizer/semantic-arc-opts-redundantcopyopts.sil
@@ -1,5 +1,7 @@
 // RUN: %target-sil-opt -module-name Swift -enable-sil-verify-all -semantic-arc-opts -sil-semantic-arc-peepholes-redundant-copyvalue-elim %s | %FileCheck %s
 
+// XFAIL: OS=linux-android, OS=linux-androideabi
+
 // NOTE: Some of our tests here depend on borrow elimination /not/ running!
 // Please do not add it to clean up the IR like we did in
 // semanticarcopts-loadcopy-to-loadborrow!
diff --git a/swift/utils/build-script-impl b/swift/utils/build-script-impl
index 16e05052609..7ab8cebfab8 100755
--- a/swift/utils/build-script-impl
+++ b/swift/utils/build-script-impl
@@ -2622,6 +2622,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     echo "Cleaning the libdispatch build directory"
                     call rm -rf "${LIBDISPATCH_BUILD_DIR}"
                   fi
+                  call ln -sf "${SWIFT_BUILD_PATH}/lib/swift" "${ANDROID_NDK}/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib"
 
                   cmake_options=(
                     -DENABLE_SWIFT=YES
diff --git a/swift/utils/swift_build_support/swift_build_support/products/swift_testing.py b/swift/utils/swift_build_support/swift_build_support/products/swift_testing.py
index 324d1a77eea..e88601a8701 100644
--- a/swift/utils/swift_build_support/swift_build_support/products/swift_testing.py
+++ b/swift/utils/swift_build_support/swift_build_support/products/swift_testing.py
@@ -13,6 +13,9 @@
 import os
 
 from build_swift.build_swift.versions import Version
+from ..host_specific_configuration \
+    import HostSpecificConfiguration
+from ..targets import StdlibDeploymentTarget
 
 from . import cmake_product
 from . import product
@@ -115,6 +117,24 @@ class SwiftTestingCMakeShim(cmake_product.CMakeProduct):
         # FIXME: If we build macros for the builder, specify the path.
         self.cmake_options.define('SwiftTesting_MACRO', 'NO')
 
+        if host_target.startswith('android') and self.is_cross_compile_target(host_target):
+            host_config = HostSpecificConfiguration(host_target, self.args)
+            self.cmake_options.extend(host_config.cmake_options)
+            triple = '%s-unknown-linux-android%s' % (self.args.android_arch,
+                                                     self.args.android_api_level)
+            flags = '-target %s ' % (triple)
+
+            flags += '-resource-dir %s/lib/swift ' % (
+                     self.host_install_destdir(host_target) + self.args.install_prefix)
+
+            ndk_path = StdlibDeploymentTarget.get_target_for_name(host_target).platform.ndk_toolchain_path(self.args)
+            flags += '-sdk %s/sysroot ' % (ndk_path)
+            flags += '-tools-directory %s/bin' % (ndk_path)
+            self.cmake_options.define('CMAKE_Swift_FLAGS', flags)
+            self.cmake_options.define('CMAKE_Swift_COMPILER_TARGET', triple)
+            self.cmake_options.define('CMAKE_CXX_COMPILER_WORKS', 'True')
+            self.cmake_options.define('CMAKE_FIND_ROOT_PATH', self.args.cross_compile_deps_path)
+
         self.generate_toolchain_file_for_darwin_or_linux(
             host_target, override_macos_deployment_version=override_deployment_version)
         self.build_with_cmake([], self.args.build_variant, [],
diff --git a/swift-corelibs-foundation/CMakeLists.txt b/swift-corelibs-foundation/CMakeLists.txt
index 7f290d16..95366592 100644
--- a/swift-corelibs-foundation/CMakeLists.txt
+++ b/swift-corelibs-foundation/CMakeLists.txt
@@ -51,6 +51,7 @@ if(NOT CMAKE_SYSTEM_NAME STREQUAL Windows)
     endif()
 endif()
 
+set(CMAKE_SHARED_LINKER_FLAGS "")
 set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
diff --git a/swift-testing/CMakeLists.txt b/swift-testing/CMakeLists.txt
index 1be9a4b..bd7b1bd 100644
--- a/swift-testing/CMakeLists.txt
+++ b/swift-testing/CMakeLists.txt
@@ -28,6 +28,7 @@ list(APPEND CMAKE_MODULE_PATH
   ${PROJECT_SOURCE_DIR}/cmake/modules
   ${PROJECT_SOURCE_DIR}/cmake/modules/shared)
 
+set(CMAKE_SHARED_LINKER_FLAGS "")
 set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
